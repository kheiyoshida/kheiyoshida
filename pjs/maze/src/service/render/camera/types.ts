
export type CameraMoveValues = {
  zDelta?: number;
  turnDelta?: number;
  upDown?: number;
};
