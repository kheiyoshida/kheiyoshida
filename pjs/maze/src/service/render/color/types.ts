import { ColorEffectPattern, ColorIntention } from '../../../domain/color/types'
import { ColorPalette } from './palette'

export type ColorManipFn = (
  palette: ColorPalette,
  params: ColorIntention
) => ColorPalette

export type ColorManipFnMap = { [k in ColorEffectPattern]: ColorManipFn }
