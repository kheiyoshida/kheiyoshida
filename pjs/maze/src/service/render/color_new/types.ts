import { Color } from 'maze-gl'

export type IColorScheme = {
  /**
   * the base color for anything unlit (background, unlit objects).
   * light calculation result adds up to this color:
   *
   * in short:
   * color = unlitColor + (material * light)
   *
   */
  readonly unlitColor: Color

  /**
   * the half part of "delta" to the final color result.
   */
  readonly lightColor: Color

  /**
   * another half part of "delta" to the final color result.
   */
  readonly materialColor: Color

  /**
   * if lighting is inverted, the flag would be true
   */
  readonly litInversion: boolean

  /**
   * rotate the scheme's hue
   * @param degrees
   */
  rotateHue(degrees: number): void

  /**
   * increase overall saturation
   */
  increaseSaturation(delta: number, max?: number): void

  /**
   * make the color range lighter or darker
   */
  moveLightnessRange(delta: number): void

  /**
   * increase the lightness of light color
   */
  increaseLitLevel(delta: number): void

  /**
   * fade out light color.
   * when fading in again, call fadeInLight()
   */
  fadeOutLight(): void
  fadeInLight(): void
}
