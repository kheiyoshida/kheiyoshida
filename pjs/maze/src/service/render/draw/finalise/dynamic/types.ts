import { Position3D } from "p5utils/src/3d"

/**
 * geometry specifications to represent `RenderModel`
 */
export type GeometrySpec = {
  coords: ShapeCoordinates
  normalPosition: Position3D
}

/**
 * rectangle coordinates
 */
export type ShapeCoordinates = Position3D[]
