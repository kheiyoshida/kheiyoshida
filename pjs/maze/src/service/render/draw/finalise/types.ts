import type { Geometry, Image } from "p5"
import { Position3D, SphericalAngles } from "p5utils/src/3d"

export type DrawableObject = {
  geometry: Geometry
  position: Position3D
  rotation?: SphericalAngles
  texture?: Image
}
