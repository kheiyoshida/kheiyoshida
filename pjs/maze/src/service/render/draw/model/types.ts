/* eslint-disable @typescript-eslint/no-unused-vars */
import type { RenderGrid } from '../../../../domain/translate/renderGrid/renderSpec'
import type { RenderBlockPosition } from '../scaffold'

export type GridConverter = (rendewrGrid: RenderGrid) => ModelCodeGrid

export type RenderModelType = 'dynamic' | 'static'

export type RenderModel = DynamicModel | StaticModel

type _RenderModel<T extends RenderModelType> = {
  type: T
  code: T extends 'dynamic' ? DynamicModelCode : StaticModelCode
  position: RenderBlockPosition
}

export type DynamicModel = _RenderModel<'dynamic'>
export type StaticModel = _RenderModel<'static'>

export type ModelCodeGrid = ModelCodeGridLayer[]

export type ModelCodeGridLayer = [CompoundModelCode, CompoundModelCode, CompoundModelCode]

export type CompoundModelCode = ModelCode[]

export type ModelCode = DynamicModelCode | StaticModelCode

export enum DynamicModelCode {
  Floor = 'Floor',
  Ceil = 'Ceil',
  SideWall = 'SideWall',
  FrontWall = 'FrontWall',
  Stair = 'Stair',
  StairCeil = 'StairCeil',
  BoxTop = 'BoxTop',
  BoxMiddle = 'BoxMiddle',
  BoxBottom = 'BoxBottom',
  BoxStair = 'BoxStair',
}

export enum StaticModelCode {
  Pole = 'Pole',
  Tile = 'Tile'
}

type Incl = {
  includes: (arg: ModelCode) => boolean
}

export const dynamicModelCodes = Object.entries(DynamicModelCode).map(
  ([_, val]) => val
) as DynamicModelCode[] & Incl

export const staticModelCodes = Object.entries(StaticModelCode).map(
  ([_, val]) => val
) as StaticModelCode[] & Incl
