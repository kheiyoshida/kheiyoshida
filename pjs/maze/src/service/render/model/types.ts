import { Position3D } from 'p5utils/src/3d'
import * as RenderSpec from '../../../domain/translate/compose/renderSpec'

export type ModelGrid = ModelGridLayer[]

export type ModelGridLayer = [CompoundRenderModel, CompoundRenderModel, CompoundRenderModel]

export type CompoundRenderModel = RenderModel[]

export enum RenderModel {
  Floor = 'Floor',
  Ceil = 'Ceil',
  SideWall = 'SideWall',
  FrontWall = 'FrontWall',
  Stair = 'Stair',
  StairCeil = 'StairCeil',
}

export type ShapeCoordinates = Position3D[]
export type GeometryCoordinates = ShapeCoordinates[]

export type RenderBlockPosition = {
  z: number
  x: RenderSpec.RenderPosition
  y?: never // maybe y in the future
}

export type RenderBlockCoords = {
  front: RenderBlockLayer
  rear: RenderBlockLayer
}

/**
 * defines the layer of render block's front or rear
 *  tl         tr
 *   '---------'
 *   |         |
 *   |         |
 *   '---------'
 *  bl         br
 */
export type RenderBlockLayer = {
  tl: Position3D
  tr: Position3D
  bl: Position3D
  br: Position3D
}
