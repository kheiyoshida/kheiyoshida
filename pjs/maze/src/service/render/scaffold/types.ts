import { Position3D } from 'p5utils/src/3d/types'
import * as RenderSpec from '../../../domain/compose/renderSpec'

export type RenderBlockPosition = {
  z: number
  x: RenderSpec.RenderPosition
  y?: never // maybe y in the future
}

export type RenderBlockCoords = {
  front: RenderBlockLayer
  rear: RenderBlockLayer
}

/**
 * defines the layer of render block's front or rear
 *  tl         tr
 *   '---------'
 *   |         |
 *   |         |
 *   '---------'
 *  bl         br
 */
export type RenderBlockLayer = {
  tl: Position3D
  tr: Position3D
  bl: Position3D
  br: Position3D
}

export type ScaffoldValues = {
  floor: number
  path: number
  wall: number
  distortionRange: number
  distortionSpeed: number
}

/**
 * collection of consistant Positon3D
 * each RenderBlock should get the concrete value from this
 * currently supports 7 layers (=RenderGrid.length + 1)
 */
export type Scaffold<T extends ScaffoldEntity = Position3D> = ScaffoldLayer<T>[]

export type ScaffoldEntity = Position3D | DistortionDelta | unknown

export type ScaffoldLayerPartKey = 'upper' | 'lower'
export type ScaffoldLayer<T extends ScaffoldEntity = Position3D> = {
  [k in ScaffoldLayerPartKey]: ScaffoldLayerPart<T>
}

/**
 * index with `ScaffoldLayerCoordPosition`
 */
export type ScaffoldLayerPart<T extends ScaffoldEntity = Position3D> = T[]

export const ScaffoldLayerPartLength = 4

export enum ScaffoldLayerCoordPosition {
  LL = 0,
  CL = 1,
  CR = 2,
  RR = 3,
}

export type ScaffoldKey = {
  layerIndex: number
  partKey: ScaffoldLayerPartKey
  position: ScaffoldLayerCoordPosition
}

export interface DistortionDelta {
  values: Position3D
  move(range: number, speed?: number): void
}