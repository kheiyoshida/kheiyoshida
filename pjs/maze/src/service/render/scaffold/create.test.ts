import {
  createScaffold,
  createScaffoldLayer,
  createScaffoldLayerPart,
  getLayerZValue,
  makeGetScaffoldXValue,
  makegetLayerYValue,
} from './create.ts'
import { ScaffoldLayerCoordPosition, ScaffoldValues } from './types.ts'

const defaultValues: ScaffoldValues = {
  floor: 400,
  path: 800,
  wall: 400,
  distortionRange: 0,
  distortionSpeed: 0,
}

test(`${createScaffold.name}`, () => {
  const scaffold = createScaffold(defaultValues)
  expect(scaffold).toHaveLength(7)
})

test(`${createScaffoldLayer.name}`, () => {
  const layer = createScaffoldLayer(1, defaultValues)
  expect(layer.lower[ScaffoldLayerCoordPosition.LL][2]).toBe(
    getLayerZValue(1, defaultValues.floor, defaultValues.path)
  )
})

test.each([
  [0, 300],
  [1, -300],
  [2, -1300],
  [3, -1900],
  [4, -2900],
])(`${getLayerZValue.name} (%i)`, (index, expected) => {
  expect(getLayerZValue(index, 600, 1000)).toBe(expected)
})

test.each([
  [400, -200],
  [800, -600],
  [1200, -1000],
])(`${makegetLayerYValue.name} (%i)`, (height, expected) => {
  const getY = makegetLayerYValue(height, defaultValues.wall)
  expect(getY('lower')).toBe(defaultValues.wall / 2)
  expect(getY('upper')).toBe(expected)
})

test(`${createScaffoldLayerPart.name}`, () => {
  const [y, z] = [500, 1500]
  const layerPart = createScaffoldLayerPart(y, z, defaultValues)
  layerPart.forEach((pos, i) => {
    expect(pos[0]).toBe(makeGetScaffoldXValue(defaultValues.floor, defaultValues.path)(i))
    expect(pos[1]).toBe(y)
    expect(pos[2]).toBe(z)
  })
})

test.each([
  [ScaffoldLayerCoordPosition.LL, -1000],
  [ScaffoldLayerCoordPosition.CL, -200],
  [ScaffoldLayerCoordPosition.CR, 200],
  [ScaffoldLayerCoordPosition.RR, 1000],
])(`${makeGetScaffoldXValue.name}(%s)`, (position, expected) => {
  const getX = makeGetScaffoldXValue(defaultValues.floor, defaultValues.path)
  expect(getX(position)).toBe(expected)
})
