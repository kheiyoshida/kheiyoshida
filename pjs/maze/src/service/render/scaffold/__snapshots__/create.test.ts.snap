// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScaffold 1`] = `
[
  {
    "lower": [
      [
        -1500,
        500,
        500,
      ],
      [
        -500,
        500,
        500,
      ],
      [
        500,
        500,
        500,
      ],
      [
        1500,
        500,
        500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        500,
      ],
      [
        -500,
        -500,
        500,
      ],
      [
        500,
        -500,
        500,
      ],
      [
        1500,
        -500,
        500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -500,
      ],
      [
        -500,
        500,
        -500,
      ],
      [
        500,
        500,
        -500,
      ],
      [
        1500,
        500,
        -500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -500,
      ],
      [
        -500,
        -500,
        -500,
      ],
      [
        500,
        -500,
        -500,
      ],
      [
        1500,
        -500,
        -500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -1500,
      ],
      [
        -500,
        500,
        -1500,
      ],
      [
        500,
        500,
        -1500,
      ],
      [
        1500,
        500,
        -1500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -1500,
      ],
      [
        -500,
        -500,
        -1500,
      ],
      [
        500,
        -500,
        -1500,
      ],
      [
        1500,
        -500,
        -1500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -2500,
      ],
      [
        -500,
        500,
        -2500,
      ],
      [
        500,
        500,
        -2500,
      ],
      [
        1500,
        500,
        -2500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -2500,
      ],
      [
        -500,
        -500,
        -2500,
      ],
      [
        500,
        -500,
        -2500,
      ],
      [
        1500,
        -500,
        -2500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -3500,
      ],
      [
        -500,
        500,
        -3500,
      ],
      [
        500,
        500,
        -3500,
      ],
      [
        1500,
        500,
        -3500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -3500,
      ],
      [
        -500,
        -500,
        -3500,
      ],
      [
        500,
        -500,
        -3500,
      ],
      [
        1500,
        -500,
        -3500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -4500,
      ],
      [
        -500,
        500,
        -4500,
      ],
      [
        500,
        500,
        -4500,
      ],
      [
        1500,
        500,
        -4500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -4500,
      ],
      [
        -500,
        -500,
        -4500,
      ],
      [
        500,
        -500,
        -4500,
      ],
      [
        1500,
        -500,
        -4500,
      ],
    ],
  },
  {
    "lower": [
      [
        -1500,
        500,
        -5500,
      ],
      [
        -500,
        500,
        -5500,
      ],
      [
        500,
        500,
        -5500,
      ],
      [
        1500,
        500,
        -5500,
      ],
    ],
    "upper": [
      [
        -1500,
        -500,
        -5500,
      ],
      [
        -500,
        -500,
        -5500,
      ],
      [
        500,
        -500,
        -5500,
      ],
      [
        1500,
        -500,
        -5500,
      ],
    ],
  },
]
`;
